vsphere/vsphere.go

```diff
@@ -149,35 +152,95 @@ func (ops *vsphereOps) Create(opts interface{}, labels map[string]string) (inter
 
 	volumeOptions.Datastore = datastore
 
-	diskBasePath := filepath.Clean(ds.Path(diskDirectory)) + "/"
-	err = ds.CreateDirectory(ctx, diskBasePath, false)
-	if err != nil && err != vclib.ErrFileAlreadyExist {
-		logrus.Errorf("Cannot create dir %#v. err %s", diskBasePath, err)
-		return nil, err
-	}
+	var (
+		trueVar = true
+		disk    diskmanagers.VirtualDisk
+	)
 
-	diskPath := diskBasePath + volumeOptions.Name + ".vmdk"
-	disk := diskmanagers.VirtualDisk{
-		DiskPath:      diskPath,
-		VolumeOptions: volumeOptions,
-	}
-
-	diskPath, err = disk.Create(ctx, ds)
+	about := vmObj.Client().ServiceContent.About
+	apiVersion, err := version.NewVersion(about.ApiVersion)
 	if err != nil {
-		logrus.Errorf("Failed to create a vsphere volume with volumeOptions: %+v on "+
-			"datastore: %s. err: %+v", volumeOptions, datastore, err)
-		return nil, err
+		return nil, fmt.Errorf("failed to detect vSphere API version due to: %v", err)
 	}
 
-	// Get the canonical path for the volume path.
-	canonicalVolumePath, err := getCanonicalVolumePath(ctx, vmObj.Datacenter, diskPath)
+	keepDiskVersion, err := version.NewVersion(keepAfterDeleteVMApiVersion)
 	if err != nil {
-		logrus.Errorf("Failed to get canonical vsphere disk path for: %s with "+
-			"volumeOptions: %+v on datastore: %s. err: %+v", diskPath, volumeOptions, datastore, err)
-		return nil, err
-	}
+		return nil, fmt.Errorf("failed to parse vSphere API version that supports keepAfterDeleteVM due to: %v", err)
+	}
+
+	if apiVersion.GreaterThan(keepDiskVersion) || apiVersion.Equal(keepDiskVersion) {
+		// create disk using the new API so it doesn't get deleted after VM deletion
+		m := vslm.NewObjectManager(vmObj.Client())
+		spec := types.VslmCreateSpec{
+			Name:              volumeOptions.Name,
+			CapacityInMB:      int64(volumeOptions.CapacityKB*1024) / units.MB,
+			KeepAfterDeleteVm: &trueVar,
+			BackingSpec: &types.VslmCreateSpecDiskFileBackingSpec{
+				VslmCreateSpecBackingSpec: types.VslmCreateSpecBackingSpec{
+					Datastore: ds.Reference(),
+				},
+				ProvisioningType: string(types.BaseConfigInfoDiskFileBackingInfoProvisioningTypeThin),
+			},
+		}
+
+		task, err := m.CreateDisk(ctx, spec)
+		if err != nil {
+			return nil, err
+		}
+
+		res, err := task.WaitForResult(ctx)
+		if err != nil {
+			return nil, err
+		}
 
-	disk.DiskPath = canonicalVolumePath
+		if res == nil {
+			return nil, fmt.Errorf("got empty result while creating vSphere disk")
+		}
+
+		myDisk := res.Result.(types.VStorageObject)
+		logrus.Infof("Created vSphere disk (VMDK) with ID: %s", myDisk.Config.Id.Id)
+
+		fileInfo, ok := myDisk.Config.Backing.(*types.BaseConfigInfoDiskFileBackingInfo)
+		if !ok {
+			return nil, fmt.Errorf("failed to get disk file path: %v", myDisk)
+		}
+		disk = diskmanagers.VirtualDisk{
+			DiskPath:      fileInfo.FilePath,
+			VolumeOptions: volumeOptions,
+		}
+	} else {
+		logrus.Warnf("Detected older vSphere version: %s. If VM gets deleted without detaching disk, disk "+
```

Comment:
Ideally this should be a warning/alert. Can we do a similar check in porx ?

Reply:
Removing it from here and will add as an error log and alert in the porx PR. Removing so that we don't have duplicate logs.